### ISSUE TITLE
3.1 - Nested prefixes don't consider parent prefixes for templates

### ISSUE THREAD

**Issue by sdustinh** [cakephp/cakephp#6293]
Consider the following...

```php
- src/
    - Controller/
        - Admin/
            - Product/
                - ProductsController.php
    - Template/
        - Admin/
            - Layout/
                - default.ctp
            
            - Product/
                - Products/
                    - view.ctp
        - Layout/
            - default.ctp
```

```php
Router::scope('/', ['plugin' => 'Products'], function ($routes) {
    $routes->prefix('Admin', function ($routes) {
        $routes->prefix('Product', function ($routes) {
            $routes->routeClass('DashedRoute');
            $routes->connect(
                '/:id',
                ['controller' => 'Products', 'action' => 'view'],
                [
                    'pass' => ['id'],
                    'id' => '[0-9]+',
                    '_name' => 'Admin.Products.viewProduct'
                ]
            );
        });
    );
});
```

It makes it very easy to space out my logic and make my application/plugins better but the parent prefix paths aren't considered when generating the paths.

```
// View::_getSubPaths();
[
    'Admin\Product\Layout\',
    'Layout\'
]
```

**Comment by markstory** [comment]
Cascading up the prefix paths could result is a few extra failed filesystem hits for each layout/template. How often, and how deeply do you generally nest prefixes?

**Comment by sdustinh** [comment]
Rarely but I just don't have the option to do it now. I tried this and it didn't work...

```php
// src/Template/Admin/Product/Layout/default.ctp
$this->extend('Products./Admin/default');
echo $this->fetch('content');
```

Even though I rarely do this now I could definitely see the benefit in deeper nesting in the future. It would make my controllers thinner and actions more specific, more reliant on components and other importable classes to share between controllers. It would force the convention on my templates too. Lots of benefits to be had.

Until a more permanent solution is figured out here is what I am using...

```php
protected function _getSubPaths($basePath)
    {
        $paths = [$basePath];
        
        if (!empty($this->request->params['prefix'])) {
            $pieces = explode('/', $this->request->params['prefix']);
            $total = count($pieces);
            
            for ($i = 1; $total >= $i; ++$i) {
                $prefixPath = array_map(
                    'Cake\Utility\Inflector::camelize',
                    array_slice($pieces, 0, $i)
                );
                array_unshift(
                    $paths,
                    implode('/', $prefixPath) . DS . $basePath
                );
            }
        }
        
        return $paths;
    }
```

It is **completely** untested outside of my limited use and already I see hiccups with it but it scales and was a 5-minute fix. Maybe a solution would be a way to define how deeply to check for nesting? Or maybe a way for me to white/blacklist paths?

```php
// In the view class...

$this->addTemplatePrefix('Products.Admin');
$this->addTemplatePath('Products./Admin/Layout');
$this->addTemplatePath('Products', false); // ignore

// Maybe just explicitly define which paths to use
// Non-prefixed always available
$this->templatePrefixPaths(['Admin.Product', 'Admin']);
```

**Comment by markstory** [comment]
If we're going to support nested prefix cascades it will be a core feature with no knobs. People who don't like it can make their own view classes :smile:

**Comment by HavokInspiration** [comment]
This is a really good idea.

**Comment by HavokInspiration** [comment]
I'm thinking about taking care of this one.
I have one question though. The issue is flagged for the 3.1 release, but can't it be for the next 3.0 release ? The original changes were introduced in 3.0.1.

**Comment by markstory** [comment]
@HavokInspiration I'm flexible either release works for me.

**Comment by markstory** [comment]
Implemented on master for 3.0.3.
